File: FpWorkshop/FSharpPortion/Step03_Sequences.fs
namespace FSharpPortion

module Step03_Sequences =

  let run () =
    // Challenge 1 --------------------------------------
    let anArray = [|1; 3; __|]

    ensure (anArray = [|1; 3; 5|])

    // Challenge 2 --------------------------------------
    ensure (anArray.[1] = __)

    // Challenge 3 --------------------------------------
    let aList = [7; 11; __; 17]

    ensure (aList = [7; 11; 13; 17])

    // Challenge 4 --------------------------------------
    let aSet = Set.ofList [1; 2; 3; __]

    ensure (aSet = Set.ofList [1; 2; 3; 7])

    // Challenge 5 --------------------------------------
    let aSeq = Seq.ofList [__; 100; 103]

    ensure (aSeq = Seq.ofList [99; 100; 103])

    // NOTE: From now on we'll use lists and the `List` module but most of
    // the following exists for the other containers we've mentioned also.
    // Seq is the most general (e.g. can be use on arrays, lists, etc) but
    // it will convert the containers to sequences.

    // Challenge 6 --------------------------------------
    // `head` gets the first element but it's unsafe because it can throw exceptions!
    ensure (List.head [] = 42)

    // Challenge 7 --------------------------------------
    // Similar story with `tail`
    ensure (List.tail [] = [2; 3; 4])

    // NOTE: because `head` and `tail` are partial function (i.e. they don't handle
    // all possible inputs) they are considered unsafe and should be avoided in
    // production code. We'll discover safer ways of dealing with lists later...

    // Challenge 8 --------------------------------------
    // `filter` picks elements that match a "check function" (aka predicate)

    let startsWithH (x: string) = x.StartsWith("h")

    let r1 = List.filter startsWithH ["hi"; "howdy"; "yo"; "sup"; "hello"]

    ensure (r1 = __)

    // Challenge 9 --------------------------------------
    // Try it with an unnamed function (hint "not equals" is `<>` in F#)
    let noYo = List.filter (fun x -> __) ["hi"; "howdy"; "yo"; "sup"; "hello"]

    let noYoExpected = ["hi"; "howdy"; "sup"; "hello"]

    ensure (noYo = noYoExpected)

    // Challenge 10 --------------------------------------
    // Try it with a partially applied function
    let noYo = List.filter (__ "yo") ["hi"; "howdy"; "yo"; "sup"; "hello"]

    ensure (noYo = noYoExpected)

    // Challenge 11 --------------------------------------
    // `map` transforms all values in a container making a new container
    let allIncremented = List.map (fun x -> __) [0; 1; 2; 3]

    ensure (allIncremented = [1; 2; 3; 4])

    // It's kinda like `map` does this
    (*
      List.map myFunction [1; 2; 3; 4; 5]
      -> [myFunction 1; myFunction 2; myFunction 3; myFunction 4; myFunction 5]
      -> results after applying :)
    *)

    // Challenge 12 --------------------------------------
    // `collect` transforms values into containers and merges them all together
    // It's a little odd at first, but you can think of it as being like "map then squish"
    // or "map then concat"
    let double x = [x; x]
    let doubled = List.collect double [1; 2; 3]

    ensure (doubled = __)

    // (SPOILER) In this case, `collect` does 
    (*
      [double 1; double 2; double 3] -> apply each ->
      [[1; 1]; [2; 2]; [3; 3]] -> squish! ->
      [1; 1; 2; 2; 3; 3]
    *)

    // Challenge 13 --------------------------------------
    // `collect` is super powerful. It can even be used to do a sort of
    // `map` and `filter` in one step
    let exclaimOnlyStartsWithH =
      ["hi"; "howdy"; "yo"; "sup"; "hello"]
        |> List.collect (fun v -> if startsWithH v then [v + "!"] else [])

    ensure (exclaimOnlyStartsWithH  = __)

    // Challenge 14 --------------------------------------
    // `fold` rolls-up the values in a collection to form a result
    let sum = List.fold (+) 0 [1; 2; 3; 4]

    ensure (sum = __)

    // You can think of this as building up an operation that looks like
    (*
      ((((0 + 1) + 2) + 3) + 4)
    *)

    // Challenge 15 --------------------------------------
    // the `0` was the initial value, or the value used if given an empty list
    let a1 = List.fold (+) 42 []
    ensure (a1 = __)

    let a2 = List.fold (+) __ [1; 2; 3]
    ensure (a2 = 106)

    // Challenge 16 --------------------------------------
    // the result type can be different from the elements' type!
    let stringLen (x: string) = x.Length

    let totalStringLen = List.fold (fun result aStr -> __) 0 ["good"; "day"; "mate"]
    ensure (totalStringLen = 11)

File: FpWorkshop/FpWorkshop/Step03_Sequences.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace FpWorkshop
{
  public class Step03_Sequences
  {
    public static void Run()
    {
      // Challenge 1 --------------------------------------
      // What's your intuition on arrays? In other words, why use it over other enumerables?
      var anArray = new[] {1, 3, ReplaceMe.__<int>()};

      Help.Ensure(anArray.SequenceEqual(new [] {1, 3, 5}));

      // Challenge 2 --------------------------------------
      Help.Ensure(anArray[2] == ReplaceMe.__<int>());

      // Challenge 3 --------------------------------------
      // What's your intuition on lists? In other words, why use it over other enumerables?
      var aList = new List<int> { 7, 11, ReplaceMe.__<int>(), 17 };

      Help.Ensure(aList.SequenceEqual(new List<int> { 7, 11, 13, 17 }));

      // Challenge 4 --------------------------------------
      // What's your intuition on sets? In other words, why use it over other enumerables?
      var aSet = new HashSet<int> {1, 2, 3, ReplaceMe.__<int>() };

      Help.Ensure(aSet.SequenceEqual(new HashSet<int> {1, 2, 3, 7 }));

      // Challenge 5 --------------------------------------
      // Seq in F# is very very similar to an IEnumerable in C#.
      // What's your intuition on IEnumerables? In other words, why use it over more specific types?
      var anEnumerable = Enumerable.Range(25, 100);

      Help.Ensure(anEnumerable.First() == ReplaceMe.__<int>());

      // NOTE: From now on we'll use IEnumerable because it's the preferred
      // type of LINQ

      // Challenge 6 --------------------------------------
      Help.Ensure(new int[] { }.First() == 42);

      // Challenge 7 --------------------------------------
      // In C# you can approximate tail with Skip 1, in F# this is the same as `Seq.drop 1`
      Help.Ensure(new int[] { }.Skip(1).SequenceEqual(new [] { 2, 3, 4 }));

      // Challenge 8 --------------------------------------
      var r1 = new[] {"hi", "howdy", "yo", "sup", "hello"}.Where(StartsWithH);

      Help.Ensure(r1.SequenceEqual(ReplaceMe.__<IEnumerable<string>>()));

      // Challenge 9 --------------------------------------
      var noYo = new[] {"hi", "howdy", "yo", "sup", "hello"}
        .Where(x => ReplaceMe.__<bool>());

      Help.Ensure(noYo.SequenceEqual(new[] { "hi", "howdy", "sup", "hello" }));

      // Challenge 10 --------------------------------------
      // Can't be done in C#

      // Challenge 11 --------------------------------------
      // `Select` is LINQ's version of `map`
      var allIncremented = new[] {0, 1, 2, 3}
        .Select(x => ReplaceMe.__<int>());

      Help.Ensure(allIncremented.SequenceEqual(new[] {1, 2, 3, 4}));

      // Challenge 12 --------------------------------------
      // `SelectMany` is C#'s version of `collect`
      var doubled = new[] {1, 2, 3}
        .SelectMany(Double);

      Help.Ensure(doubled.SequenceEqual(ReplaceMe.__<IEnumerable<int>>()));

      // Challenge 13 --------------------------------------
      // `collect` is super powerful. It can even be used to do a sort of
      // `map` and `filter` in one step
      var exclaimOnlyStartsWithH =
        new [] {"hi", "howdy", "yo", "sup", "hello" }
          .SelectMany(v => StartsWithH(v) ? new [] {v + "!"} : new string[] {});

      Help.Ensure(exclaimOnlyStartsWithH.SequenceEqual(ReplaceMe.__<IEnumerable<string>>()));

      // Challenge 14 --------------------------------------
      // `Aggregate` is the LINQ version of fold
      var sum = new[] {1, 2, 3, 4}.Aggregate(0, (a, b) => a + b);

      Help.Ensure(sum == ReplaceMe.__<int>());

      // Challenge 15 --------------------------------------
      // the `0` was the initial value, or the value used if given an empty list
      var a1 = Enumerable.Empty<int>().Aggregate(0, (a, b) => a + b);
      Help.Ensure(a1 == ReplaceMe.__<int>());

      var a2 = new [] {1, 2, 3}.Aggregate(ReplaceMe.__<int>(), (a, b) => a + b);
      Help.Ensure(a2 == 106);

      // Challenge 16 --------------------------------------
      // the result type can be different from the elements' type!
      Func<string, int> stringLen = x => x.Length;

      var totalStringLen = new[] {"good", "day", "mate"}
        .Aggregate(0, (result, aStr) => ReplaceMe.__<int>());

      Help.Ensure(totalStringLen == 11);
    }

    private static bool StartsWithH(string x) => x.StartsWith("h");
    private static IEnumerable<T> Double<T>(T x) => new[] {x, x};
  }
}
